File: services/server/src/safe_socket/safe_socket.py
import socket

def recv_all(socket: socket.socket, size):
    left_to_read = size
    bytes_total = b''

    while left_to_read:
        bytes = socket.recv(left_to_read)

        if not bytes:
            break

        left_to_read -= len(bytes)
        bytes_total += bytes

    return bytes_total


def send_all(socket: socket.socket, bytes):
    written = 0

    while written < len(bytes):
        n = socket.send(bytes[written:])
        written += n

    return written

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

func SendAll(socket io.Writer, bytes []byte) error {
	written := 0

	for written < len(bytes) {
		n, err := socket.Write(bytes[written:])
		written += n

		if err != nil {
			return err
		}
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	leftToRead := size
	buffTotal := make([]byte, 0, size)

	for leftToRead > 0 {
		buff := make([]byte, leftToRead)
		n, err := socket.Read(buff)
		leftToRead -= n
		buffTotal = append(buffTotal, buff[:n]...)

		if err == io.EOF {
			break
		} else if err != nil {
			return nil, err
		}
	}

	return buffTotal, nil
}
